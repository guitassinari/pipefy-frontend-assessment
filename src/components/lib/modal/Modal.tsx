import { HTMLAttributes } from 'hoist-non-react-statics/node_modules/@types/react'
import React from 'react'
import { ModalCard, ModalContent, ModalActionsFooter } from './Modal.styles'

interface ModalProps {
  primaryAction?: () => any
  customPrimaryActionText?: string
}

const Modal: React.FC<ModalProps & HTMLAttributes<HTMLDivElement>> = ({ children, primaryAction, customPrimaryActionText}) => (
  <ModalCard>
    <ModalContent>
      {children}
    </ModalContent>
    <ModalActionsFooter>
      {primaryAction && (
        <button onClick={() => primaryAction()} role="button">
          {customPrimaryActionText || 'Close'}
        </button>
      )}
    </ModalActionsFooter>
  </ModalCard>
)

export default Modal