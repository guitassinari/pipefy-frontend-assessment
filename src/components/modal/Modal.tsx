import React, { HTMLAttributes } from 'react'
import ReactDOM from 'react-dom'
import {default as DesignSystemModal} from 'components/lib/modal'
import { MODAL_PORTAL_ID } from 'components/modal-portal'
import { Backdrop } from './Modal.styles'

interface ModalProps {
  primaryAction?: () => any
  customPrimaryActionText?: string
}

const Modal: React.FC<ModalProps & HTMLAttributes<HTMLDivElement>> = ({ children, ...rest }) => {
  const container = document.getElementById(MODAL_PORTAL_ID)

  if(!container) {
    return null
  }

  const content = (
    <Backdrop>
      <DesignSystemModal {...rest}>
        {children}
      </DesignSystemModal>
    </Backdrop>
  )
  
  return ReactDOM.createPortal(content, container as any)
}
export default Modal