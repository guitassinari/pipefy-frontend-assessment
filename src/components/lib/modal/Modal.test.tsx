import React from 'react';
import { render, fireEvent } from '@testing-library/react'
import Modal from '.';

describe('Modal', () => {
  it('matches snapshot', () => {
    const component = render(<Modal>Modal test</Modal>)

    expect(component.asFragment()).toMatchSnapshot()
  })

  it('renders its received children', () => {
    const component = render(<Modal>Modal test</Modal>)

    expect(component.getByText('Modal test')).toBeInTheDocument()
  })

  describe('when receives a primaryAction', () => {
    it('renders the primary action button with a default text', () => {
      const component = render(<Modal primaryAction={jest.fn()}>Modal test</Modal>)

      expect(component.getByRole('button', { name: 'Close' })).toBeInTheDocument()
    })

    describe('when the primary button is clicked', () => {
      it('executes the primaryAction callback', () => {
        const primaryActionCallback = jest.fn()
        const component = render(<Modal primaryAction={primaryActionCallback}>Modal test</Modal>)

        const primaryButton = component.getByRole('button', { name: 'Close' })
        fireEvent.click(primaryButton)

        expect(primaryActionCallback).toHaveBeenCalled()
      })
    })

    describe('whenr receives a customPrimaryActionText', () => {
      it('displays the button with tue custom text', () => {
        const customText = 'Custom'
        const component = render(
          <Modal primaryAction={jest.fn()} customPrimaryActionText={customText}>
            Modal test
          </Modal>
        )

        expect(component.getByRole('button', { name: customText })).toBeInTheDocument()
      })
    })
  })
})
