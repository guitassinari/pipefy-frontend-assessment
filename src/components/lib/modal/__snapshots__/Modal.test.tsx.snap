// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL ffOZty"
  >
    <div
      class="sc-gsDKAQ bucHQF"
    >
      Modal test
    </div>
    <div
      class="sc-dkPtRN eGuCUl"
    />
  </div>
</DocumentFragment>
`;
